lib/linked_list.rb

```diff
@@ -18,71 +18,158 @@ def initialize
+
+      until temp.nil?
+      
+        if temp.data == value 
+          return value
+        end
+      
+        temp = temp.next
+      end
+
+      return false
     end
 
     # method to return the max value in the linked list
     # returns the data value and not the node
     def find_max
-      raise NotImplementedError
+      current = @head 
+      until current.nil?
+
+        if (max.nil? || current.data > max)
+          max = current.data
+        end
+        
+        current = current.next
+      end
+
+      return max
     end
 
     # method to return the min value in the linked list
     # returns the data value and not the node
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(n)
+    # Space Complexity: O(n)
     def find_min
-      raise NotImplementedError
+      min = nil
+      current = @head
+
+      until current.nil?
+
+        if (min.nil? || current.data < min)
+          min = current.data
+        end
+
+        current = current.next
+      end
+
+      return min
     end
 
 
     # method that returns the length of the singly linked list
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(n)
+    # Space Complexity: O(1) because this list isn't growing over time
     def length
-      raise NotImplementedError
+      length = 0
+    if !@head
+      return length
+    else
+      current = @head
+      while current
+        length += 1
+        current = current.next
+      end
+    end
+    return length
     end
 
     # method that returns the value at a given index in the linked list
     # index count starts at 0
     # returns nil if there are fewer nodes in the linked list than the index value
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(n)
+    # Space Complexity: O(1)
     def get_at_index(index)
-      raise NotImplementedError
+      # return nil unless @head
+      if @head == nil
+        return nil
+      end
+
+      current = @head
+      index.times do 
+        # return nil unless current.next
+        if (current.next == nil) #check if index is out of bounds
+          return nil
+        end
+        current = current.next
+      end
+
+      return current.data
     end
 
     # method to print all the values in the linked list
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(n)
+    # Space Complexity: O(n)
     def visit
-      raise NotImplementedError
+      print_all_values = nil 
+      current = @head
+      
+      until current.nil?
+        print current.data
+        current = current.next
+      end
+
     end
 
     # method to delete the first node found with specified value
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(n)
+    # Space Complexity: O(n)
     def delete(value)
-      raise NotImplementedError
+      if @head.nil?
+        return nil 
+      end
+
+      previous = @head 
+      current = @head 
+

```

Comment:
You also need to check to see if value is at the head of the list

```suggestion
      if @head.value == value
        @head = @head.next
        return
      end
```